### DISCUSSION TITLE
Is PackageDependency for VCLibs in appxmanifest required?

### DISCUSSION THREAD

**Discussion by aries-zhang** [microsoft/WindowsAppSDK#4886] [Q&A]
In the "prerequisit" section of this doc [Windows App SDK deployment guide for framework-dependent packaged apps](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/deploy-packaged-apps#prerequisites), it states 
> For packaged apps, the VCLibs framework package dependency is a requirement. For more info, see [C++ Runtime framework packages for Desktop Bridge](https://learn.microsoft.com/en-us/troubleshoot/cpp/c-runtime-packages-desktop-bridge).

In [C++ Runtime framework packages for Desktop Bridge](https://learn.microsoft.com/en-us/troubleshoot/cpp/c-runtime-packages-desktop-bridge), it says to reference the VCLibs package, the following code should be added in `appxmanifest` file.
```
<Dependencies>
    <PackageDependency Name="Microsoft.VCLibs.140.00.UWPDesktop" MinVersion="14.0.24217.0" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" />
</Dependencies>
```

However, with the above code, it seems my app package is no longer able to be installed:
```
Add-AppxPackage : Deployment failed with HRESULT: 0x80073CF3, Package failed updates, dependency or 
conflict validation. 
Windows cannot install package <package-name> because this package depends on a 
framework that could not be found. Provide the framework "Microsoft.VCLibs.140.00.UWPDesktop" 
published by "CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US", 
with neutral or x64 processor architecture and minimum version 14.0.33728.0, 
along with this package to install. The frameworks with name "Microsoft.VCLibs.140.00.UWPDesktop" 
currently installed are: {}
NOTE: For additional information, look for [ActivityId] d255bb11-2bf8-0000-e145-f7d2f82bdb01 in 
the Event Log or use the command line 
Get-AppPackageLog -ActivityID d255bb11-2bf8-0000-e145-f7d2f82bdb01 
At line:1 char:1
+ Add-AppxPackage .\<package-name> ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : WriteError: (C:\Users\Aries\....msixbundle:String) [Add-AppxPackage], IOException
+ FullyQualifiedErrorId : DeploymentError,Microsoft.Windows.Appx.PackageManager.Commands.AddAppxPackageCommand
```

Removing this line fixes the error. This [issue](https://github.com/microsoft/microsoft-ui-xaml/issues/1542) also mentioned removing the line fixes the deployment error.

Also noticed that the app package does not seem to have any difference with or without the VCLibs PackageDependency in appxmanifest.

Does anyone know if this declaration in appxmanifest is required?

**Comment by whiskhub** [comment]
I am also not quite sure if it is a real requirement. At least it also works without it on first sight.

To be able to install the app with the PackageDependency, you must first install the dependency package that you can find at `%ProgramFiles(x86)%\Microsoft SDKs\Windows Kits\10\ExtensionSDKs\Microsoft.VCLibs.Desktop\14.0`. Or you use the App Installer GUI, which will download and install the dependency automatically.

**Comment by DarranRowe** [comment]
If you also notice, it also states that .NET 6 or later is required.

From a .NET side, this may seem obvious, but this would imply that a C++ or other language application using the Windows App SDK would also require .NET. So it is always difficult to take these statements as is.

Anyway, that isn't the issue here. When you are using Add-AppxPackage to install a package, you really should be using the command line options to specify the locations of the dependencies. Installing via Add-AppxPackage doesn't automatically query the store for dependencies so you have to supply them yourself. You can only get away with not supplying a dependency if it is already installed on that system.

Also, the Windows App SDK itself is built with the "Hybrid CRT" build method. So it technically doesn't need the VC runtime. However, it is quite easy to use a package that isn't built this way. So it is possible to remove the reference if you are sure that nothing in your application will ever need the runtime, but you should be careful since without the VC runtime package, the application can look at the system version of the VC redistributable, and so everything will seemingly work correctly until one day where it just fails.

**Comment by Gaoyifei1011** [comment]
> In the "prerequisit" section of this doc [Windows App SDK deployment guide for framework-dependent packaged apps](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/deploy-packaged-apps#prerequisites), it states
> 
> > For packaged apps, the VCLibs framework package dependency is a requirement. For more info, see [C++ Runtime framework packages for Desktop Bridge](https://learn.microsoft.com/en-us/troubleshoot/cpp/c-runtime-packages-desktop-bridge).
> 
> In [C++ Runtime framework packages for Desktop Bridge](https://learn.microsoft.com/en-us/troubleshoot/cpp/c-runtime-packages-desktop-bridge), it says to reference the VCLibs package, the following code should be added in `appxmanifest` file.
> 
> ```
> <Dependencies>
>     <PackageDependency Name="Microsoft.VCLibs.140.00.UWPDesktop" MinVersion="14.0.24217.0" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" />
> </Dependencies>
> ```
> 
> However, with the above code, it seems my app package is no longer able to be installed:
> 
> ```
> Add-AppxPackage : Deployment failed with HRESULT: 0x80073CF3, Package failed updates, dependency or 
> conflict validation. 
> Windows cannot install package <package-name> because this package depends on a 
> framework that could not be found. Provide the framework "Microsoft.VCLibs.140.00.UWPDesktop" 
> published by "CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US", 
> with neutral or x64 processor architecture and minimum version 14.0.33728.0, 
> along with this package to install. The frameworks with name "Microsoft.VCLibs.140.00.UWPDesktop" 
> currently installed are: {}
> NOTE: For additional information, look for [ActivityId] d255bb11-2bf8-0000-e145-f7d2f82bdb01 in 
> the Event Log or use the command line 
> Get-AppPackageLog -ActivityID d255bb11-2bf8-0000-e145-f7d2f82bdb01 
> At line:1 char:1
> + Add-AppxPackage .\<package-name> ...
> + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> + CategoryInfo          : WriteError: (C:\Users\Aries\....msixbundle:String) [Add-AppxPackage], IOException
> + FullyQualifiedErrorId : DeploymentError,Microsoft.Windows.Appx.PackageManager.Commands.AddAppxPackageCommand
> ```
> 
> Removing this line fixes the error. This [issue](https://github.com/microsoft/microsoft-ui-xaml/issues/1542) also mentioned removing the line fixes the deployment error.
> 
> Also noticed that the app package does not seem to have any difference with or without the VCLibs PackageDependency in appxmanifest.
> 
> Does anyone know if this declaration in appxmanifest is required?

你可以在 wapproj 文件中设置 ``` <WinUISDKReferences>false</WinUISDKReferences> ``` 去检查是否需要此依赖，当打包的应用添加了此属性后，如果不需要此依赖，它会自动删除。

You can check whether this dependency is required by setting ``` <WinUISDKReferences>false</WinUISDKReferences> ``` in the wapproj file. Once the packaged application has added this property, if the dependency is not required, it will be automatically removed.

**Comment by lauren-ciha** [comment]
Thank you for reaching out! The VCLibs dependency was required in projects that existed before Hybrid CRT was introduced in [version 1.0.3](https://github.com/microsoft/WindowsAppSDK/blob/73613b4b83df39b1858b978c7956dd1315a2e784/docs/Coding-Guidelines/HybridCRT.md#hybrid-crt). Any projects with newer versions of the WindowsAppSDK can set `<WinUISDKReferences>` to false. We've [updated the templates](https://github.com/microsoft/WindowsAppSDK/pull/5635) to reflect this. You can learn more about Hybrid CRT [here](https://github.com/microsoft/WindowsAppSDK/blob/73613b4b83df39b1858b978c7956dd1315a2e784/docs/Coding-Guidelines/HybridCRT.md).
